docs/assertions_guide.md

```diff
@@ -0,0 +1,315 @@
+
+```c++
+class List {
+private:
+  // Lists contain nodes, connected in a chain using pointers
+  struct Node {
+    Node *next;
+    Node *prev;
+    int datum;
+  };
+
+  // Member variables
+  Node *first;
+  Node *last;
+  int num_nodes;
+
+  // Check represenation invariants with assert
+  void check_invariants() {
+    // Either it's empty with null first/last pointers,
+    // or it's not empty and neither should be null
+    assert(!first && !last || first && last);
+
+    // There's nothing before the first node
+    // and nothing after the last node.
+    assert(first->prev == nullptr);
+    assert(last->next == nullptr);
+
+    // Manually counting the number of nodes matches
+    // the number we're tracking in num_nodes
+    int count = 0;
+    for(Node *n = first; n; n = n->next) {
+      ++count;
+    }
+    assert(count == num_nodes);
+  }
+
+public:
+
+  // Default constructor. Creates an empty list.
+  List : first(nullptr), last(nullptr), num_nodes(0) {
+    check_invariants(); // ensure our list is set up
+  }
+
+  //EFFECTS:  inserts datum into the front of the list
+  void push_front(const T &datum) {
+
+    // A bunch of complicated code that rearranges
+    // pointers to add a node to the list
+    // ...
+
+    check_invariants(); // make sure we didn't screw up the pointers
+  }
+};
+```
+{: data-highlight="15-33,39,49" }
+
+## Debug Failed Assertions
+
+Let's say you're running `Image_public_test.exe` on project 2 and
+one of your `Matrix_at()` assertions fails:
+
+```console
+$ ./Image_public_tests
+Running test: test_image_basic
+Image_public_test.exe: Matrix.cpp:72: int* Matrix_at(Matrix*, int, int): Assertion `0 <= row && row < Matrix_height(mat)' failed.
+Aborted
+```
+{: data-highlight="16,39,49" data-variant="legacy" }
+
+That's nice, but just a line number (i.e. `Matrix.cpp:72`) doesn't really give us much to go on:
+
+<img src="images/assertions010.png" width="480px" />
+
+We need more information. Let's run the code in our **visual debugger** instead. (You don't need to add breakpoints, just run it.)
+
+It should pause when the `assert` fails so that we can inspect the state of the program and figure out why the assertion is failing. First, check the local variables:
+
+<img src="images/assertions020.png" width="768px" />
+
+It looks like the problem is that the `column` parameter is too large (given width 5, only 0-4 column indices are allowed). We can figure out where this is coming from by clicking on the the stack frame for the calling context, which is the `Matrix_fill()` function:
+
+<img src="images/assertions030.png" width="768px" />
+
+Now we can actually see the call to `Matrix_at()` that broke its REQUIRES clause. In this implementation of `Matrix_fill()`, the `c` variable is allowed to reach value `5`, which is one too large. We need to adjust the conditions on the loops to fix this.
+
+Because we had assertions in `Matrix_at()`, it warned us right away that `Matrix_fill()` was up to no good!
+
```

Comment:
Consider adding an `Acknowledgements` section with a CC license similar to other tutorials.

Reply:
done!